packages/navigator/src/browser/open-editors-widget/navigator-open-editors-tree-model.ts

```diff
@@ -0,0 +1,218 @@
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct } from '@theia/core/shared/inversify';
+import { FileNode, FileStatNode, FileTreeModel } from '@theia/filesystem/lib/browser';
+import {
+    ApplicationShell,
+    CompositeTreeNode,
+    open,
+    NavigatableWidget,
+    OpenerService,
+    SelectableTreeNode,
+    TreeNode,
+    Widget,
+    ExpandableTreeNode
+} from '@theia/core/lib/browser';
+import { WorkspaceService } from '@theia/workspace/lib/browser';
+import debounce = require('@theia/core/shared/lodash.debounce');
+import { DisposableCollection } from '@theia/core/lib/common';
+
+export interface OpenEditorNode extends FileStatNode {
+    widget: Widget;
+};
+
+export namespace OpenEditorNode {
+    export function is(node: object | undefined): node is OpenEditorNode {
+        return FileStatNode.is(node) && 'widget' in node;
+    }
+}
+
+@injectable()
+export class OpenEditorsModel extends FileTreeModel {
+    static GROUP_NODE_ID_PREFIX = 'group-node';
+    static AREA_NODE_ID_PREFIX = 'area-node';
+
+    @inject(ApplicationShell) protected readonly applicationShell: ApplicationShell;
+    @inject(WorkspaceService) protected readonly workspaceService: WorkspaceService;
+    @inject(OpenerService) protected readonly openerService: OpenerService;
+
+    protected toDisposeOnPreviewWidgetReplaced = new DisposableCollection();
+
+    protected _editorWidgetsByArea = new Map<ApplicationShell.Area, NavigatableWidget[]>();
+
+    get editorWidgets(): NavigatableWidget[] {
+        let editorWidgets: NavigatableWidget[] = [];
+        this._editorWidgetsByArea.forEach(widgets => {
+            editorWidgets = [...editorWidgets, ...widgets];
+        });
+        return editorWidgets;
+    }
+
+    getEditorWidgetsByGroup(id: ApplicationShell.Area | number): NavigatableWidget[] | undefined {
+        const idAsNum = Number(id);
+        if (!isNaN(idAsNum)) {
+            return this._editorWidgetsByGroup.get(idAsNum);
+        }
+        return this._editorWidgetsByArea.get(id as ApplicationShell.Area);
+    }
+
+    protected _editorWidgetsByGroup = new Map<number, NavigatableWidget[]>();
+
+    @postConstruct()
+    protected init(): void {
+        super.init();
+        this.setupHandlers();
+        this.initializeRoot();
+    }
+
+    protected setupHandlers(): void {
+        this.toDispose.push(this.selectionService.onSelectionChanged(selection => {
+            const { widget } = (selection[0] as OpenEditorNode);
+            this.applicationShell.activateWidget(widget.id);
+        }));
```

Comment:
It looks like this is basically intended to be a click handler `handleClickEvent`? If so, one option would be to override the click handler in the tree widget, as is done in the `navigator-widget` code.

Testing this also made me notice that it doesn't look like selection is reflected in the tree. I only see hover marking, not selected / active marking.

Reply:
Good call, I've used the tree widget's selection method to do the activation. Even though the regular NavigatorTree uses "revealWidget" to activate the widget, I've gone activateWidget, since there's no obvious visual indication of which widget is being selected, especially when multiple groups are open. Also, I'm relying on the shell's knowledge of `activeWidget` to execute some of context menu commands reliably